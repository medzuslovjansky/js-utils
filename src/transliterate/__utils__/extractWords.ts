#!/usr/bin/env node

import fs from 'node:fs';
import _ from 'lodash';

function* extractWords(str: string) {
  // include letters and combining marks
  const regex = /([\p{L}\p{M}]+)/gu;
  let match;

  while ((match = regex.exec(str)) !== null) {
    yield match[1];
  }
}

function* extractWordsFromFile(filePath: string) {
  const raw = fs.readFileSync(filePath, 'utf8');
  yield* extractWords(raw);
}

function* readAllWords() {
  yield* extractWordsFromFile(
    'src/adjective/__snapshots__/declensionAdjective.test.ts.snap',
  );
  yield* extractWordsFromFile(
    'src/noun/__snapshots__/declensionNoun.test.ts.snap',
  );
  yield* extractWordsFromFile('src/numeral/testCases.json');
  yield* extractWordsFromFile('src/pronoun/testCases.json');
  yield* extractWordsFromFile('src/verb/testCases.json');
  yield* extractWordsFromFile('scripts/dictionary.txt');
}

export default function () {
  const allWords = [...readAllWords()];
  const allUpper = allWords.filter((word) => word[0].toUpperCase() === word[0]);
  const allLower = allWords.filter((word) => word[0].toLowerCase() === word[0]);
  const allUniqueWords = _.uniq([...allLower, ...allUpper]);
  return bucketSortByLength(allUniqueWords);
}

function bucketSortByLength(data: string[]): string[] {
  const buckets: string[][] = [];

  for (const word of data) {
    const length = word.length;
    if (!buckets[length]) {
      buckets[length] = [];
    }
    buckets[length].push(word);
  }

  return buckets.flat();
}
