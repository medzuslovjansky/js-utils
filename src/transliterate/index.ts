import {
  transliterate as _transliterate,
  FlavorizationType,
  TransliterationType,
} from './transliterate';

import { FlavorisationBCP47Code } from '../constants';

export default function transliterate(
  text: string,
  lang: FlavorisationBCP47Code,
  preprocessed = false,
): string {
  switch (lang) {
    case 'isv-Latn':
      return _transliterate(
        text,
        TransliterationType.Latin,
        FlavorizationType.Standard,
        preprocessed,
      );
    case 'isv-Cyrl':
      return _transliterate(
        text,
        TransliterationType.StandardCyrillic,
        FlavorizationType.Standard,
        preprocessed,
      );
    case 'isv-Glag':
      return _transliterate(
        text,
        TransliterationType.Glagolitic,
        FlavorizationType.Standard,
        preprocessed,
      );
    case 'isv-x-fonipa':
      return _transliterate(
        text,
        TransliterationType.IPA,
        FlavorizationType.Etymological,
        preprocessed,
      );
    case 'isv-Latn-x-etymolog':
      return _transliterate(
        text,
        TransliterationType.Latin,
        FlavorizationType.Etymological,
        preprocessed,
      );
    case 'isv-Cyrl-x-etymolog':
      return _transliterate(
        text,
        TransliterationType.StandardCyrillic,
        FlavorizationType.Etymological,
        preprocessed,
      );
    case 'isv-Glag-x-etymolog':
      return _transliterate(
        text,
        TransliterationType.Glagolitic,
        FlavorizationType.Etymological,
        preprocessed,
      );
    case 'isv-Cyrl-x-iotated':
      return _transliterate(
        text,
        TransliterationType.TraditionalIotatedCyrillic,
        FlavorizationType.Standard,
        preprocessed,
      );
    case 'isv-Cyrl-x-iotated-ext':
      return _transliterate(
        text,
        TransliterationType.TraditionalIotatedCyrillic,
        FlavorizationType.CyrillicExtended,
        preprocessed,
      );
    case 'isv-Cyrl-x-northern':
      return _transliterate(
        text,
        TransliterationType.StandardCyrillic,
        FlavorizationType.Northern,
        preprocessed,
      );
    case 'isv-Cyrl-x-sloviant':
      return _transliterate(
        text,
        TransliterationType.StandardCyrillic,
        FlavorizationType.Slovianto,
        preprocessed,
      );
    case 'isv-Cyrl-x-southern':
      return _transliterate(
        text,
        TransliterationType.StandardCyrillic,
        FlavorizationType.Southern,
        preprocessed,
      );
    case 'isv-Latn-PL':
      return _transliterate(
        text,
        TransliterationType.Polish,
        FlavorizationType.Etymological,
        preprocessed,
      );
    case 'isv-Latn-x-ascii':
      return _transliterate(
        text,
        TransliterationType.ASCII,
        FlavorizationType.Standard,
        preprocessed,
      );
    case 'isv-Latn-x-northern':
      return _transliterate(
        text,
        TransliterationType.Latin,
        FlavorizationType.Northern,
        preprocessed,
      );
    case 'isv-Latn-x-sloviant':
      return _transliterate(
        text,
        TransliterationType.Latin,
        FlavorizationType.Slovianto,
        preprocessed,
      );
    case 'isv-Latn-x-southern':
      return _transliterate(
        text,
        TransliterationType.Latin,
        FlavorizationType.Southern,
        preprocessed,
      );
    case 'isv-Glag-x-northern':
      return _transliterate(
        text,
        TransliterationType.Glagolitic,
        FlavorizationType.Northern,
        preprocessed,
      );
    case 'isv-Glag-x-southern':
      return _transliterate(
        text,
        TransliterationType.Glagolitic,
        FlavorizationType.Southern,
        preprocessed,
      );
    case 'isv-Glag-x-sloviant':
      return _transliterate(
        text,
        TransliterationType.Glagolitic,
        FlavorizationType.Slovianto,
        preprocessed,
      );
    case 'isv':
      return text;
    default:
      throw new TypeError(`Unsupported IETF BCP47 tag: ${lang}`);
  }
}
