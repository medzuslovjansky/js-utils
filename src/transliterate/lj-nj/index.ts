import ljeExceptions from './exceptions-lj.json';
import njeExceptions from './exceptions-nj.json';
import njeEndings from './endings-nje.json';
import {
  ENDS,
  MISMATCH,
  findTrieWord,
  findTriePosition,
  WHOLE,
} from './findTrieWord';

/**
 * Check whether we should soften lj to ĺj
 */
export function ljeCheck(word: string) {
  return findTrieWord(word, ljeExceptions) === WHOLE;
}

/**
 * Check whether we should soften lj to ĺj
 */
export function ljePosition(word: string) {
  return word.lastIndexOf('lj');
}

/**
 * Check whether we should soften nj to ńj
 */
export function njeCheck(word: string) {
  return (
    findTrieWord(word, njeEndings) === ENDS &&
    findTrieWord(word, njeExceptions) === MISMATCH
  );
}

export function njePosition(word: string): number {
  return findTriePosition(njeEndings, word);
}
