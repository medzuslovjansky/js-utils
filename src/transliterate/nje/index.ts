import type { Dict } from './Dict';
import njeExceptions from './exceptions';
import rules from './rules';

const ENDS = 1;
const WHOLE = 0;
const MISMATCH = -1;

function findTrieWord(word: string, dict: Dict): Dict | number | undefined {
  let node: Dict | number = dict;
  const length = word.length;
  let chr = '';
  let i: number;

  for (i = length - 1; typeof node === 'object' && i >= 0; i--) {
    chr = word[i];
    node = node[chr];
  }

  return node === 0 ? (i === -1 ? WHOLE : ENDS) : MISMATCH;
}

/**
 * Check whether we should soften nj to Å„j
 */
export function njeCheck(word: string) {
  return (
    findTrieWord(word, rules) === ENDS &&
    findTrieWord(word, njeExceptions) === MISMATCH
  );
}

export function njePosition(word: string): number {
  const length = word.length;
  let node: Dict | number = rules;
  let i: number;
  let chr: string;
  for (i = length - 1; typeof node === 'object' && i >= 0; i--) {
    chr = word[i];
    node = node[chr];
  }

  return node === 0 ? i + 1 : -1;
}
