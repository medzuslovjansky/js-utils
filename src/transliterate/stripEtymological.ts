const MAP: Record<string, string> = {
  å: 'a',
  ć: 'č',
  ď: 'd',
  đ: 'dž',
  è: 'e',
  ę: 'e',
  ĺ: 'l',
  ń: 'n',
  ȯ: 'o',
  ŕ: 'r',
  ś: 's',
  ť: 't',
  ų: 'u',
  ŭ: 'v',
  ź: 'z',
};

export function stripEtymological(word: string): string {
  let result = '';
  for (const char of word) {
    result += MAP[char] || char;
  }
  return result;
}
