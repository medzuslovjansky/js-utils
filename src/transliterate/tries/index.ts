export { default as ljj } from './ljj';
export { default as nj } from './nj';
export { default as njjOverride } from './njj-override';
