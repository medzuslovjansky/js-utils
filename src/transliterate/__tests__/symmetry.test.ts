import * as fixtures from '../../__utils__/fixtures';
import transliterate from '..';

describe('transliteration integrity', () => {
  test('symmetry between Latin and Cyrillic', () => {
    const exceptions = fixtures
      .word_forms()
      .reduce((acc: string[], word: string) => {
        const lat = transliterate(word, 'isv-Latn', true);
        const cyr = transliterate(word, 'isv-Cyrl', true);
        const lat2 = transliterate(cyr, 'isv-Latn');
        const cyr2 = transliterate(lat2, 'isv-Cyrl', true);

        if (lat.includes('lj') || lat.includes('nj')) {
          return cyr === cyr2
            ? acc
            : [...acc, `${word} (${lat}) -> ${cyr} vs. ${cyr2}`];
        }

        return acc;
      }, []);

    expect(exceptions).toMatchSnapshot();
  });
});
