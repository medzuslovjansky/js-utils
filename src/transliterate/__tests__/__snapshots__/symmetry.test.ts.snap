// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transliteration integrity symmetry between Latin and Cyrillic 1`] = `[]`;
